daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/unparsers/runtime1/LayeredSequenceUnparser.scala

```diff
@@ -67,36 +68,52 @@ class LayeredSequenceUnparser(
-    // finished, the layerDOS content will be written to the
-    // layerUnderlyingDOS, which will subsequently be finished. Like above, it
-    // is important to pass an immutable UState into the setFinished function
-    // because that state is stored in the DOS (as finishedFormatInfo) and its
-    // use may be delayed to after the actual UState has been changed. The
-    // UState has almost certainly changed since the last cloneForSuspension
-    // call, so we need a new clone to finish this DOS.
-    val layerDOSLast = layerDOS.lastInChain
-    val formatInfoPost = state.asInstanceOf[UStateMain].cloneForSuspension(layerDOSLast)
-    layerDOSLast.setFinished(formatInfoPost)
-
-    // clean up resources - note however, that due to suspensions, the whole
-    // layer stack is potentially still needed, so not clear what can be
-    // cleaned up at this point.
-    layerTransformer.removeLayer(layerDOS, state)
-
-    // reset the state so subsequent unparsers write to the following DOS
-    state.dataOutputStream = layerFollowingDOS
+    try {
+      val layerDriver = LayerDriver(formatInfoPre, ctxt.layerRuntimeData)
+
+      // New layerDOS is where the layer will unparse into. Ultimately anything written
+      // to layerDOS ends up, post transform, in layerUnderlyingDOS
+      val layerDOS = layerDriver.addOutputLayer(layerUnderlyingDOS)
+
+      // unparse the layer body into layerDOS
+      state.dataOutputStream = layerDOS
+      super.unparse(state)
+      // now we're done unparsing the layer recursively.
+      // While doing that unparsing, the data output stream may have been split, so the
+      // DOS in the state may no longer be the layerDOS.
+      //
+      // However, it is when whatever DOS is in the state at this point, that, when that
+      // DOS is consolidated and written out, that is when the layer is finished
+      // and the wrap-up of the layer (such as writing output variables) can occur.
+      //
+      val endOfLayerUnparseDOS = state.dataOutputStream
+      val formatInfoPost =
+        state.asInstanceOf[UStateMain].cloneForSuspension(endOfLayerUnparseDOS)
+
+      // setFinished on this end-of-layer-unparse data-output-stream  ensures
+      // that the layerDOS gets close() called on it.
+      endOfLayerUnparseDOS.setFinished(formatInfoPost)
+
+      // clean up resources - note however, that due to suspensions, the whole
+      // layer stack is potentially still needed, so
+      // nothing can be cleaned up at this point.
+    } catch {
+      case u: UnsuppressableException =>
+        throw u
+      case lre: LayerRuntimeException =>
+        throw lre
+      case re: RuntimeException =>
+        throw new LayerRuntimeException(re)
+      case pe: UnparseError =>
+        throw pe
+      case sde: RuntimeSchemaDefinitionError =>
+        throw sde
+      case e: Exception =>
+        state.toss(state.toProcessingError(new LayerUnexpectedException(e)))
```

Comment:
I don't think so. A layer can fail in many ways that are hard to anticipate due to data problems. Data problems can always be due to speculative parsing down the wrong branch. 

Users can choose to catch errors and make things RSDE if they want, but if we impose that most Exceptions are fatal then users will just have to wrap lots of try/catches around their code to anticipate and convert all the ordinary "things that can go wrong due to bad data" and convert them into processing errors. 

I've taken one consistent position here which I think enables good composition properties for schemas that use layers. Exceptions become PE. RuntimeExceptions are fatal, as are Error and Daffodil inteernal errors which are UnsuppressableException/Abort. 

Sometimes layer authors will want to catch specific RuntimeExceptions and convert them to PE. Our built in layers do this for code that converts encoding names to charsets. If that fails, it's a PE, though the charset findByName method throws RuntimeException.

Reply:
Sounds reasonable. Maybe one small suggestion then is to rename LayerUnexpectedException to something else? When I see "unexpected" all I think is it's a bug that someone forgot to handle. If non-RuntimeExceptions are allowed behavior then I wouldn't considered them unexpected. I would only consider RuntimeExceptions unexpected since layers shouldn't be throwing those.

And if not already, please make sure to document in the Layer API that RuntimeExceptions are fatal and other Exceptions are treated as PE.

Reply:
Hmmm. Should it just be LayerException? 

LayerException and LayerRuntimeException then have the same relationship as Exception and RuntimeException in terms of being routine vs. special.

Reply:
Seems reasonable to me. It we want to be extra clear, we could do something like LayerProcessingException and LayerFatalException, but I'm not sure the extra verbosity adds much extra if it's understood that RuntimeExceptions are fatal and others aren't, and it's not as 1-to-1 with Exception vs RuntimeException.